integrated_channels/lms_utils.py

```diff
@@ -42,14 +56,44 @@ def get_course_certificate(course_id, user):
 def get_single_user_grade(course_id, user):
     """
     Returns a grade for the user (must be a django.contrib.auth.User instance).
+    If there is a problem loading the CourseGradeFactory class, throws NotConnectedToOpenEdX
+
     Args:
         course_key (CourseLocator): The course to retrieve user grades for.
 
     Returns:
         A serializable list of grade responses
     """
     if not CourseGradeFactory:
-        return None
+        raise NotConnectedToOpenEdX(
+            'To use this function, this package must be '
+            'installed in an Open edX environment.'
+        )
     course_key = CourseKey.from_string(course_id)
-    course_grade = CourseGradeFactory().read(user.username, course_key=course_key)
+    course_grade = CourseGradeFactory().read(user, course_key=course_key)
     return course_grade
+
+
+def get_course_details(course_id):
+    """
+    Returns:
```

Comment:
Again, this is not a clean interface. You just moved the problems: you are mixing "Return an error code OR throw an exception, the same 20 line function."  That is 2 different problems the caller now has to handle. Either ALWAYS return none+error code or ALWAYS let the exceptions pass through.

Reply:
To explain: the problem with this is that now the calling code will have to write _extra_ code and do _more_ weird error handling, not to mention the code author will have to go look inside this function to figure out how to write the error handler for it.

Reply:
Ah I see what you are saying, so the current usage looks like this with a mixed bag of None vs exceptions

```
course_details = get_course_details(course_run_id)
if course_details is None:
    # Course not found, so we have nothing to report.
```

Instead we can go:

```
try:
    course_details = get_course_details(course_run_id)
except InvalidKeyError:                             # Course not found, log error
except CourseOverview.DoesNotExist:   # Course key is invalid, log error
```

This will also make this function consistent with the other ones in the pr

Reply:
note, the caller has to still handle these two exceptions, since we call this function in a loop of enrollments, we want to log and continue the loop in these two cases. The NotConnectedException is different, it should be treated as runtime exception since something is really bad, no point in continuing. But rather than swallow these exceptions to is better regardless, to propagate them from the get_course_details function.

Reply:
@edx-abolger sorry for the multiple rounds here, but this new pattern look better and more consistent hopefully